Make the Spectator button only appear in HMD (also add a debug mode for devs)

zfox23 opened the pull request:
Does what it says on the tin. Also disables the camera when switching from HMD to Desktop.

**Test plan:**
1. Don't touch the `spectatorCamera.js` script. Open Interface in Desktop mode.
2. Verify that the "SPECTATOR" HUD button is _present_ (because the `showSpectatorInDesktop` debug flag is set to `true` in this PR by default)
3. Switch to HMD mode
4. Verify that the "SPECTATOR" Tablet app is _present_
5. Quit Interface
6. Open `spectatorCamera.js` in your favorite text editor
7. Find the `showSpectatorInDesktop` variable and set it to `false`.
8. Open Interface in HMD mode
9. Verify that the "SPECTATOR" Tablet app is _present_
10. Switch to Desktop mode
11. Verify that the "SPECTATOR" HUD button is _not present_
12. Close Interface
13. Re-open Interface in Desktop mode
14. Verify that the "SPECTATOR" HUD button is _not present_
15. Switch to HMD mode
16. Verify that the "SPECTATOR" Tablet app is _present_
17. Turn the camera ON using the Tablet app. _Switch the "MONITOR SHOWS" slider to "Camera View"._
18. Switch to Desktop mode.
19. Verify that the camera entity is gone _and that your monitor correctly shows your avatar's point of view._

zfox23 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [do we disable in desktop, or not show the app at all?](https://trello.com/c/3YHEmrFk/191-do-we-disable-in-desktop-or-not-show-the-app-at-all)

zfox23 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [HMD-only](https://trello.com/c/ctI9SGWm/172-hmd-only)

CainFoool commented:
Will test.

YuppyPlays commented:
Will test once built :)

zfox23 commented:
@CainFoool Thank you for being so quick! However, because this has the Discovery label, you don't need to test it - I believe that PRs with this label should only be tested by the Discovery team. I will confirm this with the team.

CainFoool commented:
@zfox: I've only started seeing the Discovery tag recently. I was not aware of this. Let me know what they say! :)

zfox23 commented:
Okay, @CainFoool! I go back on what I said. Go ahead and QA this one.

We've decided that it's a judgement call each PR. This one is straightforward enough for external testers to do.

Thank you!

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/737/HighFidelity-Beta-PR10697-7788c1d1b9792458a778f90038bbc56f2c26c227.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/737/HighFidelity-Beta-PR10697-7788c1d1b9792458a778f90038bbc56f2c26c227.exe

CainFoool commented:
# Test Failed 

- Verify that the "SPECTATOR" HUD button is present (because the showSpectatorInDesktop debug flag is set to true in this PR by default) - *Pass*
- Switch to HMD mode - *Pass*
- Verify that the "SPECTATOR" Tablet app is present - *Fail <--*
- Quit Interface - *Pass*
- Open spectatorCamera.js in your favorite text editor - *Pass*
- Find the showSpectatorInDesktop variable and set it to false. - *Pass*
- Open Interface in HMD mode - *Pass*
- Verify that the "SPECTATOR" Tablet app is present - *Pass*
- Switch to Desktop mode - *Pass*
- Verify that the "SPECTATOR" HUD button is not present - *Pass*
- Close Interface - *Pass*
- Re-open Interface in Desktop mode - *Pass*
- Verify that the "SPECTATOR" HUD button is not present - *Pass*
- Switch to HMD mode - *Pass*
- Verify that the "SPECTATOR" Tablet app is present - *Pass*
- Turn the camera ON using the Tablet app. - *Pass*
- Switch to Desktop mode. - *Pass*
- Verify that the camera entity is gone. - *Pass*

On a side note, the Spectator camera is a pretty awesome utility. Just the one fail which needs to be fixed.

tapalisa commented:
@CainFoool Can you please run through a smoke test to make sure the new functionality hasn't broken other features? The "Core Functionality" tab is sufficient.

CainFoool commented:
@tapalisa: Will perform a smoke test shortly.

howard-stearns commented:
So, the principle functionality of this, per discussion with @zfox23 is:
1. When a user is wearing an HMD, the "SPECTATOR" app button (tablet or toolbar) _will be_ visible.
2. When a user is in Desktop mode, the "SPECTATOR" app button (tablet or toolbar) _will not be_ visible.
3. If the Spectator Camera is enabled, turn it off when switching to Desktop mode.

We need to also specify (and include in test cases), what happens with the main window display. E.g,., you have the camera on and "monitor shows" switch set to "camera view". Now go to desktop (either with tablet, the mouse on the Interface menu bar, or ctrl-1. What does the display do?

CainFoool commented:
# Smoke Test

It was a good call for Alisa to ask me to test this out. I had plenty of issues which I could replicate in regards to models/textures loading with this. 

Usually I start my interface up by right clicking on the Icon and clicking 'Go Home' and that will bring me into my domain, and that worked fine which is where I conducted the original testing. 

This time, per smoke test, I decided to start teleporting around and visiting other domains and this is where I was starting to have issues.

Take note that the Tutorial Garden is on the Same Domain, I teleported to it through the Portal and ended up hovering in mid air due to the fact that barely any of the models and textures were refusing to load. There was no textures or models processing/pending in the Stats screen.

So thinking this was a one off, I decided to restart my game and see what would happen this time. My Home Domain loaded as normal and I teleported back to the Garden. This time, most of the models had loaded but were refusing to load their full texture. (See snap I took below)
![snap1](https://user-images.githubusercontent.com/898687/27191172-ed5881dc-51ee-11e7-9d41-153137ec52b0.PNG)

I can't remember completely from the back of my head but in the back of the Tutorial Room there were some models which were hanging up on the wall, most of these didn't actually render until I came up close to them, and when I looked away I was able to see them disappear, I had to get right up close to them in order to see them. I was able to look other ways and see everything else but these ones in particular were refusing to load correctly - these ones however, appeared to have loaded their texture just fine. 

So for a third time, I restarted my game and went back to Tutorial Garden - this time, I fell right through the floor and started endlessly falling, however I caught myself and flew back up and landed and found that everything appeared to load normally, including the Skybox.

This is where I decided to AGAIN, restart my game and head to dev-welcome. This was fun. Some of the nearby models were loading (I could see their shapes) but their textures were no where to be found. I saw them pending in the Stats bar but again nothing was downloading, so I waited a minute or two to see if there would be any change and there wasn't so I went back to my Home and saw that 
 the pending count got stuck even though I teleported to my Home and nothing else would load - I was stuck on the black/blue Skybox of death with nothing around me.
![image](https://user-images.githubusercontent.com/898687/27191401-98c96ee6-51ef-11e7-9a1b-571369c14457.png)

I will attach all of my log files from these sessions below:

- [hifi-log.txt](https://github.com/highfidelity/hifi/files/1078296/hifi-log.txt)
- [hifi-log_192.168.1.231_2017-06-15_15.28.27.txt](https://github.com/highfidelity/hifi/files/1078297/hifi-log_192.168.1.231_2017-06-15_15.28.27.txt)
- [hifi-log_192.168.1.231_2017-06-15_15.32.29.txt](https://github.com/highfidelity/hifi/files/1078300/hifi-log_192.168.1.231_2017-06-15_15.32.29.txt)
- [hifi-log_192.168.1.231_2017-06-15_15.43.13.txt](https://github.com/highfidelity/hifi/files/1078301/hifi-log_192.168.1.231_2017-06-15_15.43.13.txt)
- [hifi-log_192.168.1.231_2017-06-15_15.54.58.txt](https://github.com/highfidelity/hifi/files/1078298/hifi-log_192.168.1.231_2017-06-15_15.54.58.txt)
- [hifi-log_192.168.1.231_2017-06-15_15.56.09.txt](https://github.com/highfidelity/hifi/files/1078299/hifi-log_192.168.1.231_2017-06-15_15.56.09.txt)
- [hifi-log_192.168.1.231_2017-06-15_17.09.17.txt](https://github.com/highfidelity/hifi/files/1078302/hifi-log_192.168.1.231_2017-06-15_17.09.17.txt)
- [hifi-log_192.168.1.231_2017-06-15_17.11.14.txt](https://github.com/highfidelity/hifi/files/1078304/hifi-log_192.168.1.231_2017-06-15_17.11.14.txt)
- [hifi-log_192.168.1.231_2017-06-15_17.20.08.txt](https://github.com/highfidelity/hifi/files/1078303/hifi-log_192.168.1.231_2017-06-15_17.20.08.txt)

zfox23 commented:
@CainFoool - I modified steps 17 and 19 in the test plan, and also fixed the failure you mentioned. Could you please go back and run through the test plan again? Thank you :)

davidkelly commented on scripts/system/spectatorCamera.js:
```diff
@@ -186,34 +187,60 @@
+    var button = false;
     var buttonName = "SPECTATOR";
     var tablet = null;
+    var showSpectatorInDesktop = true;
```
I'm guessing we will make this false later

zfox23 replied:
Correct!

davidkelly replied:
👍

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/754/HighFidelity-Beta-PR10697-68662bb209f1cc92da2616676883d4a0d9545da8.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/754/HighFidelity-Beta-PR10697-68662bb209f1cc92da2616676883d4a0d9545da8.exe

CainFoool commented:
@zfox23: Will do - testing now :)

CainFoool commented:
# PR Passed 

Proceeding the steps below, this PR and the required testing steps were all a pass. Great job!

- [x] Don't touch the spectatorCamera.js script. Open Interface in Desktop mode.
- [x] Verify that the "SPECTATOR" HUD button is present (because the showSpectatorInDesktop debug flag is set to true in this PR by default)
- [x] Switch to HMD mode
- [x] Verify that the "SPECTATOR" Tablet app is present
- [x] Quit Interface
- [x] Open spectatorCamera.js in your favorite text editor
- [x] Find the showSpectatorInDesktop variable and set it to false.
- [x] Open Interface in HMD mode
- [x] Verify that the "SPECTATOR" Tablet app is present
- [x] Switch to Desktop mode
- [x] Verify that the "SPECTATOR" HUD button is not present
- [x] Close Interface
- [x] Re-open Interface in Desktop mode
- [x] Verify that the "SPECTATOR" HUD button is not present
- [x] Switch to HMD mode
- [x] Verify that the "SPECTATOR" Tablet app is present
- [x] Turn the camera ON using the Tablet app. Switch the "MONITOR SHOWS" slider to "Camera View".
- [x] Switch to Desktop mode.
- [x] Verify that the camera entity is gone and that your monitor correctly shows your avatar's point of view.
